### PR TITLE
[LBSE] 'mask' / 'clipPath' (SVG) DOM changes should never trigger relayouts

### PR DESCRIPTION
**Pull request by nikolaszimmermann**
#### 26d12c80d8df13b1005c00fc45cb88ad622e2f64
<pre>
[LBSE] &apos;mask&apos; / &apos;clipPath&apos; (SVG) DOM changes should never trigger relayouts
<a href="https://bugs.webkit.org/show_bug.cgi?id=267732">https://bugs.webkit.org/show_bug.cgi?id=267732</a>

Reviewed by Rob Buis and Simon Fraser.

When clipPahUnits change via (SVG) DOM, SVGClipPathElement::svgAttributeChanged()
will trigger an async re-layout of the associated RenderSVGResourceClipper. This is
useless, as it does not repaint the elements making use of the resource. Furthermore
unnecessary work is spent in layout - avoid all of that, and repaint the old and new
boundaries of all resource users, with the help of RenderLayers repaintRect tracking,
that assures the objects is repainted at its old and new position, not leaving visual
artifacts.

Covered by existing tests in LBSE + added a new test svg/repaint/clip-path-change-unit-space.html.

* LayoutTests/platform/ios/svg/repaint/clip-path-object-bounding-box-transformed-expected.txt:
* LayoutTests/platform/ios/svg/repaint/mask-object-bounding-box-transformed-expected.txt:
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/animations/svg-element-attribute-changed-crash-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/transforms/svg-css-transforms-clip-path-expected.txt: Added.
* LayoutTests/svg/repaint/clip-path-change-unit-space-expected.txt: Added.
* LayoutTests/svg/repaint/clip-path-change-unit-space.html: Added.
* LayoutTests/svg/repaint/clip-path-object-bounding-box-transformed-expected.txt:
* LayoutTests/svg/repaint/clip-path-user-space-on-use-transformed-expected.txt:
* LayoutTests/svg/repaint/mask-object-bounding-box-transformed-expected.txt:
* LayoutTests/svg/repaint/mask-user-space-on-use-transformed-expected.txt:
* Source/WebCore/rendering/ReferencedSVGResources.cpp:
(WebCore::CSSSVGResourceElementClient::resourceChanged):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::recursiveUpdateLayerPositions):
* Source/WebCore/rendering/svg/SVGContainerLayout.cpp:
(WebCore::SVGContainerLayout::layoutChildren):
* Source/WebCore/svg/SVGClipPathElement.cpp:
(WebCore::SVGClipPathElement::svgAttributeChanged):
(WebCore::SVGClipPathElement::childrenChanged):
* Source/WebCore/svg/SVGMaskElement.cpp:
(WebCore::SVGMaskElement::svgAttributeChanged):
(WebCore::SVGMaskElement::childrenChanged):

Canonical link: <a href="https://commits.webkit.org/273293@main">https://commits.webkit.org/273293@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/6760c4721ef00f743b818bab2d3469f1bcc9e595

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/34902 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/13756 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/36945 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/37616 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/31504 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/36051 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/16155 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/10836 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/30440 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/35435 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/11656 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/31090 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/10193 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/10261 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/31178 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/38870 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/31719 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/31505 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/36287 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/10364 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/8275 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/34265 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/12180 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/30821 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8018 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/10906 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/11243 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/rendering/RenderLayer.cpp [hunk 1]
```diff
@@ -991,6 +991,19 @@ void RenderLayer::recursiveUpdateLayerPositions(OptionSet<UpdateLayerPositionsFl
...
 
+#if ENABLE(LAYER_BASED_SVG_ENGINE)
+    if (renderer().isSVGLayerAwareRenderer() && renderer().document().settings().layerBasedSVGEngineEnabled()) {
+        if (!is<RenderSVGRoot>(renderer())) {
+            ASSERT(!renderer().isFixedPositioned());
+            m_repaintStatus = RepaintStatus::NeedsFullRepaint;
+        }
+
+        // Only the outermost <svg> and / <foreignObject> are potentially scrollable.
+        if (is<RenderSVGModelObject>(renderer()) || is<RenderSVGText>(renderer()) || is<RenderSVGInline>(renderer()))
```
Comment (smfr) [inline Source/WebCore/rendering/RenderLayer.cpp]: This should use `ASSERT_IMPLIES()`

**Review by smfr — APPROVED** [review]

File: Source/WebCore/rendering/RenderLayer.cpp (hunk 1 above)
Comment (nikolaszimmermann) [reply Source/WebCore/rendering/RenderLayer.cpp]: Ok.

File: Source/WebCore/svg/SVGMaskElement.cpp [hunk 2]
```diff
@@ -128,6 +136,14 @@ void SVGMaskElement::childrenChanged(const ChildChange& change)
     if (change.source == ChildChange::Source::Parser)
         return;
 
+#if ENABLE(LAYER_BASED_SVG_ENGINE)
+    if (document().settings().layerBasedSVGEngineEnabled()) {
+        if (auto* renderer = this->renderer())
+            renderer->repaintClientsOfReferencedSVGResources();
+        return;
+    }
+#endif
```
Comment (rwlbuis) [inline Source/WebCore/svg/SVGMaskElement.cpp:145]: Some helper method may be convenient if we start using this everywhere for resources.

**Review by rwlbuis — APPROVED** [review]
LGTM too (minus the test failures)

File: Source/WebCore/svg/SVGMaskElement.cpp [hunk 3]
```diff
@@ -114,6 +114,14 @@ void SVGMaskElement::svgAttributeChanged(const QualifiedName& attrName)
     }
 
     if (PropertyRegistry::isKnownAttribute(attrName)) {
+#if ENABLE(LAYER_BASED_SVG_ENGINE)
+        if (document().settings().layerBasedSVGEngineEnabled()) {
+            if (auto* renderer = this->renderer())
```
Comment (nikolaszimmermann) [inline Source/WebCore/svg/SVGMaskElement.cpp]: Forgot to use CheckedPtr here, will fix.

Comment (nikolaszimmermann) [reply Source/WebCore/svg/SVGMaskElement.cpp]: Fixed.

File: Source/WebCore/svg/SVGMaskElement.cpp (hunk 2 above)
Comment (nikolaszimmermann) [reply Source/WebCore/svg/SVGMaskElement.cpp:145]: Yeah, we can think about a more central helper once the full pattern has emerged for more than just one or two resources.

File: Source/WebCore/rendering/svg/SVGContainerLayout.cpp [hunk 4]
```diff
@@ -75,19 +76,20 @@ void SVGContainerLayout::layoutChildren(bool containerNeedsLayout)
...
+                if (CheckedPtr shape = dynamicDowncast<RenderSVGShape>(child)) {
+                    shape->setNeedsShapeUpdate();
                     needsLayout = true;
-                }
+                } else if (CheckedPtr svgText = dynamicDowncast<RenderSVGText>(child)) {
+                    svgText->setNeedsTextMetricsUpdate();
+                    svgText->setNeedsPositioningValuesUpdate();
+                    needsLayout = true;
+                } else if (CheckedPtr resource = dynamicDowncast<RenderSVGResourceGradient>(child))
+                    resource->invalidateGradient();
```
Comment (nikolaszimmermann) [inline Source/WebCore/rendering/svg/SVGContainerLayout.cpp:91]: This change was necessary to fix `LayoutTests/svg/custom/relative-sized-content-with-resources.xhtml` - if a gradient uses relative lengths for e.g. cx/cy/fx/etc. (for radial gradients) it needs to recompute the internal `Gradient` and repaint all clients. In LBSE we can easily solve that during `SVGContainerLayout::layoutChildren` - without resorting to artificial relayout, as the legacy engine did.

File: Source/WebCore/rendering/ReferencedSVGResources.cpp [hunk 5]
```diff
@@ -60,8 +60,39 @@ WTF_MAKE_ISO_ALLOCATED_IMPL(CSSSVGResourceElementClient);
...
+        return;
+
+#if ENABLE(LAYER_BASED_SVG_ENGINE)
+    auto useUpdateLayerPositionsLogic = [&]() -> std::optional<CheckedPtr<RenderLayer>> {
+        auto& document = m_clientRenderer.document();
+        if (!document.settings().layerBasedSVGEngineEnabled())
+            return std::nullopt;
+
+        // Don't attempt to update anything during layout - the post-layout phase will invoke RenderLayer::updateLayerPosition(), if necessary.
+        if (document.view()->layoutContext().isInLayout())
```
Comment (nikolaszimmermann) [inline Source/WebCore/rendering/ReferencedSVGResources.cpp:73]: This is new: make sure we only use the `updateLayerPositions()` logic if it's safe to do so: not during layout in general, and not if the `m_clientRenderer` needs layout().

During layout repainting of old/new boundaries is already handled using `LayoutRepainter` - no need to handle it here (and would be wrong to do so, leading to debug assertions).

**Comment by nikolaszimmermann** [comment]
@smfr / @rwlbuis - please check again, I had to make sure we don't invoke RenderLayer::updateLayerPositions() neither during layout, nor if the corresponding renderers needs layout. This is covered by running LBSE tests using a debug build.

EWS did show failures, because LBSE-only logic was executed for legacy as well - fixed.

**Comment by nikolaszimmermann** [comment]
Needed to add update two iOS specific results -- patch is now ready.

File: Source/WebCore/rendering/ReferencedSVGResources.cpp [hunk 6]
```diff
@@ -60,8 +60,39 @@ WTF_MAKE_ISO_ALLOCATED_IMPL(CSSSVGResourceElementClient);
...
+
+        // If no layers are available, always use the renderer based repaint() logic.
+        if (!m_clientRenderer.hasLayer())
+            return std::nullopt;
+
+        // Use the cheaper update mechanism for all SVG renderers -- in proper subtrees, that do not need layout themselves.
+        if (!m_clientRenderer.isSVGLayerAwareRenderer() || m_clientRenderer.needsLayout())
+            return std::nullopt;
+
+        return std::make_optional(downcast<RenderLayerModelObject>(m_clientRenderer).checkedLayer());
```
Comment (rwlbuis) [inline Source/WebCore/rendering/ReferencedSVGResources.cpp:84]: I think the explicit case may not be needed.

File: Source/WebCore/rendering/ReferencedSVGResources.cpp [hunk 7]
```diff
@@ -60,8 +60,39 @@ WTF_MAKE_ISO_ALLOCATED_IMPL(CSSSVGResourceElementClient);
...
+    };
+
+    // LBSE: Instead of repainting the current boundaries, utilize RenderLayer::updateLayerPositionsAfterStyleChange() to repaint
+    // the old and the new repaint boundaries, if they differ -- instead of just the new boundaries.
+    if (auto layer = useUpdateLayerPositionsLogic()) {
+        (*layer.value()).updateLayerPositionsAfterStyleChange();
```
Comment (rwlbuis) [inline Source/WebCore/rendering/ReferencedSVGResources.cpp:90]: Is the value() part really needed?

File: Source/WebCore/rendering/ReferencedSVGResources.cpp [hunk 8]
```diff
@@ -60,8 +60,39 @@ WTF_MAKE_ISO_ALLOCATED_IMPL(CSSSVGResourceElementClient);
...
+        // Use the cheaper update mechanism for all SVG renderers -- in proper subtrees, that do not need layout themselves.
+        if (!m_clientRenderer.isSVGLayerAwareRenderer() || m_clientRenderer.needsLayout())
+            return std::nullopt;
+
+        return std::make_optional(downcast<RenderLayerModelObject>(m_clientRenderer).checkedLayer());
+    };
+
+    // LBSE: Instead of repainting the current boundaries, utilize RenderLayer::updateLayerPositionsAfterStyleChange() to repaint
+    // the old and the new repaint boundaries, if they differ -- instead of just the new boundaries.
+    if (auto layer = useUpdateLayerPositionsLogic()) {
```
Comment (rwlbuis) [inline Source/WebCore/rendering/ReferencedSVGResources.cpp:89]: Not too fond of the name, how about something like shouldUpdateLayerPositions?

File: Source/WebCore/rendering/svg/SVGContainerLayout.cpp (hunk 4 above)
Comment (rwlbuis) [reply Source/WebCore/rendering/svg/SVGContainerLayout.cpp:91]: It seem the layout and repaint phases get intermingled this way, but I guess LayoutRepainter is a bit similar and maybe there are other cases like this....

**Review by rwlbuis — COMMENTED** [review]
Mostly looks good, just some questions left.

**Pull request merged**